File: agenkit/adapters/llm/openai_compatible.py
"""
OpenAI-Compatible adapter for self-hosted and local inference services.

⚠️ IMPORTANT: When to use OpenAILLM vs OpenAICompatibleLLM ⚠️

USE OpenAILLM (agenkit.adapters.llm.openai) WHEN:
  ✅ Using the official OpenAI API (api.openai.com)
  ✅ You want GPT-4, GPT-4 Turbo, GPT-3.5, o1, o3, etc.
  ✅ You need premium features (vision, function calling, JSON mode, embeddings)
  ✅ You want official OpenAI support and SLAs
  ✅ Pay-per-token pricing is acceptable

USE OpenAICompatibleLLM (THIS MODULE) WHEN:
  ✅ Running self-hosted/local inference (vLLM, llama.cpp, SGLang, etc.)
  ✅ You want to use open models (Llama, Mistral, Qwen, Yi, etc.)
  ✅ Cost reduction is important (local inference = no API costs)
  ✅ Data privacy is required (on-premises deployment)
  ✅ You need low latency (local = no network round-trip to OpenAI)
  ✅ You want to experiment with different models easily

📖 KEY DIFFERENCE:
  - OpenAILLM: Uses https://api.openai.com/v1 (hardcoded, official OpenAI)
  - OpenAICompatibleLLM: Uses YOUR base_url (configurable, self-hosted)

  Both use the same OpenAI SDK internally, so the API is IDENTICAL.

🔧 COMPARISON EXAMPLE:

    # Official OpenAI API - for production cloud deployments
    from agenkit.adapters.llm import OpenAILLM
    openai = OpenAILLM(
        api_key="sk-...",              # Your OpenAI API key
        model="gpt-4-turbo"             # Official OpenAI model
    )
    response = await openai.complete(messages)  # Charges your OpenAI account

    # Self-hosted vLLM - for cost-effective local inference
    from agenkit.adapters.llm import OpenAICompatibleLLM
    vllm = OpenAICompatibleLLM(
        base_url="http://localhost:8000/v1",    # Your local server
        model="meta-llama/Llama-2-70b-chat-hf", # Open source model
        provider="vllm"                          # Optional metadata
    )
    response = await vllm.complete(messages)     # Free (runs on your hardware)

    # BOTH have identical APIs - just swap the import!

📚 SUPPORTED SERVICES (All OpenAI-compatible):

1. vLLM - Production-grade serving (https://docs.vllm.ai)
   Best for: High throughput, production deployments, GPU clusters

2. llama.cpp - Efficient CPU/GPU inference (https://github.com/ggerganov/llama.cpp)
   Best for: Local development, edge deployment, CPU inference, GGUF models

3. SGLang - Fast multi-turn conversations (https://sgl-project.github.io)
   Best for: Complex workflows, 29-64% faster than vLLM for some cases

4. TensorRT-LLM - NVIDIA GPU optimization (https://github.com/NVIDIA/TensorRT-LLM)
   Best for: Maximum GPU performance on NVIDIA hardware

5. Ollama - Simple local model management (https://ollama.ai)
   Best for: Quick local experimentation, one-command model downloads

6. Text Generation Inference (TGI) - Hugging Face (https://github.com/huggingface/text-generation-inference)
   Best for: Production serving with Hugging Face models

7. LM Studio - Desktop GUI for local models (https://lmstudio.ai)
   Best for: Non-technical users, visual model management

8. LocalAI - Drop-in OpenAI replacement (https://localai.io)
   Best for: OpenAI API compatibility testing, multi-backend support

9. OpenLLM - LLM serving framework (https://github.com/bentoml/OpenLLM)
   Best for: Production serving with BentoML ecosystem

EXAMPLE - vLLM local server:
    >>> from agenkit.adapters.llm import OpenAICompatibleLLM
    >>> llm = OpenAICompatibleLLM(
    ...     base_url="http://localhost:8000/v1",
    ...     model="meta-llama/Llama-2-7b-chat-hf",
    ...     provider="vllm"
    ... )
    >>> messages = [Message(role="user", content="Hello!")]
    >>> response = await llm.complete(messages)

EXAMPLE - llama.cpp server:
    >>> llm = OpenAICompatibleLLM(
    ...     base_url="http://localhost:8080/v1",
    ...     model="llama-2-7b-chat.Q4_K_M.gguf",
    ...     provider="llamacpp"
    ... )

See examples/adapters/openai_compatible/ for setup guides for each service.
"""

from collections.abc import AsyncIterator
from typing import Any

from agenkit.adapters.llm.base import LLM
from agenkit.interfaces import Message

try:
    from openai import AsyncOpenAI
except ImportError as e:
    raise ImportError(
        "The openai package is required to use OpenAICompatibleLLM. "
        "Install it with: pip install openai>=1.50.0"
    ) from e


class OpenAICompatibleLLM(LLM):
    """
    Generic adapter for OpenAI-compatible inference services.

    This adapter enables Agenkit to work with any service implementing the
    OpenAI Chat Completions API by wrapping the AsyncOpenAI SDK with a custom
    base_url parameter. This provides a consistent interface across different
    local and self-hosted inference engines.

    Args:
        base_url: Base URL of the inference service (e.g., "http://localhost:8000/v1").
            Must include the /v1 suffix for most services.
        model: Model name/identifier used by the inference service. Format varies
            by service (e.g., "meta-llama/Llama-2-7b-chat-hf" for vLLM,
            "llama-2-7b-chat" for llama.cpp).
        api_key: Optional API key. Most local services don't require authentication,
            so this defaults to "not-needed" if not provided.
        provider: Optional provider name for metadata and debugging (e.g., "vllm",
            "llamacpp", "sglang"). Helps identify which service is being used.
        timeout: Request timeout in seconds (default: 60.0). Increase for larger
            models or slower hardware.
        **client_kwargs: Additional arguments passed to AsyncOpenAI client for
            advanced configuration (e.g., max_retries, http_client).

    Raises:
        ImportError: If openai package is not installed

    Example - vLLM local deployment:
        >>> llm = OpenAICompatibleLLM(
        ...     base_url="http://localhost:8000/v1",
        ...     model="meta-llama/Llama-2-7b-chat-hf",
        ...     provider="vllm"
        ... )
        >>> response = await llm.complete([
        ...     Message(role="user", content="What is machine learning?")
        ... ])

    Example - llama.cpp server:
        >>> llm = OpenAICompatibleLLM(
        ...     base_url="http://localhost:8080/v1",
        ...     model="llama-2-7b-chat",
        ...     provider="llamacpp"
        ... )

    Example - SGLang with custom timeout:
        >>> llm = OpenAICompatibleLLM(
        ...     base_url="http://localhost:30000/v1",
        ...     model="meta-llama/Llama-2-13b-chat-hf",
        ...     provider="sglang",
        ...     timeout=120.0  # Larger model needs more time
        ... )

    Example - TensorRT-LLM with API key:
        >>> llm = OpenAICompatibleLLM(
        ...     base_url="https://api.example.com/v1",
        ...     model="llama-2-70b",
        ...     api_key="your-api-key",
        ...     provider="tensorrt"
        ... )

    Note:
        The response metadata includes both the provider name and base_url
        to help with debugging and monitoring in production deployments.
    """

    def __init__(
        self,
        base_url: str,
        model: str,
        api_key: str | None = None,
        provider: str | None = None,
        timeout: float = 60.0,
        **client_kwargs: Any,
    ) -> None:
        """Initialize OpenAI-compatible LLM adapter."""
        # Many local services don't require authentication
        if api_key is None:
            api_key = "not-needed"

        self._client = AsyncOpenAI(
            base_url=base_url,
            api_key=api_key,
            timeout=timeout,
            **client_kwargs,
        )
        self._model = model
        self._provider = provider
        self._base_url = base_url

    @property
    def model(self) -> str:
        """Return the model identifier."""
        return self._model

    async def complete(
        self,
        messages: list[Message],
        temperature: float = 1.0,
        max_tokens: int | None = None,
        **kwargs: Any,
    ) -> Message:
        """
        Generate a completion from the OpenAI-compatible service.

        Args:
            messages: Conversation history as Agenkit Messages
            temperature: Sampling temperature (0.0-2.0). Lower values make output
                more deterministic, higher values more creative.
            max_tokens: Maximum tokens to generate. None uses service default.
            **kwargs: Additional service-specific options (top_p, frequency_penalty,
                presence_penalty, stop, etc.). Passed directly to the service.

        Returns:
            Response as Agenkit Message with metadata including:
            - model: Model identifier used
            - usage: Token counts (prompt_tokens, completion_tokens, total_tokens)
            - finish_reason: Why generation stopped (stop, length, etc.)
            - provider: Provider name if specified during initialization
            - base_url: Service URL for debugging
            - id: Response ID from the service

        Raises:
            openai.APIError: For service-specific errors (connection, timeout, etc.)

        Example:
            >>> messages = [
            ...     Message(role="system", content="You are a helpful assistant."),
            ...     Message(role="user", content="What is 2+2?")
            ... ]
            >>> response = await llm.complete(messages, temperature=0.2)
            >>> print(response.content)
            >>> print(f"Tokens used: {response.metadata['usage']['total_tokens']}")
            >>> print(f"Provider: {response.metadata['provider']}")

        Example - streaming alternative:
            >>> # For real-time responses, use stream() instead
            >>> async for chunk in llm.stream(messages):
            ...     print(chunk.content, end="", flush=True)
        """
        # Validate parameters
        self._validate_llm_params(temperature, max_tokens, **kwargs)

        # Convert Agenkit Messages to OpenAI format
        openai_messages = self._convert_messages(messages)

        # Call OpenAI-compatible API
        response = await self._client.chat.completions.create(
            model=self._model,
            messages=openai_messages,
            temperature=temperature,
            max_tokens=max_tokens,
            **kwargs,
        )

        # Convert response to Agenkit Message with provider metadata
        return Message(
            role="agent",
            content=response.choices[0].message.content or "",
            metadata={
                "model": response.model,
                "usage": {
                    "prompt_tokens": response.usage.prompt_tokens if response.usage else 0,
                    "completion_tokens": response.usage.completion_tokens if response.usage else 0,
                    "total_tokens": response.usage.total_tokens if response.usage else 0,
                },
                "finish_reason": response.choices[0].finish_reason,
                "provider": self._provider or "openai_compatible",
                "base_url": self._base_url,
                "id": response.id,
            },
        )

    async def stream(
        self,
        messages: list[Message],
        temperature: float = 1.0,
        max_tokens: int | None = None,
        **kwargs: Any,
    ) -> AsyncIterator[Message]:
        """
        Stream completion chunks from the OpenAI-compatible service.

        This method streams response chunks as they're generated by the service,
        enabling real-time display and lower perceived latency for users.

        Args:
            messages: Conversation history as Agenkit Messages
            temperature: Sampling temperature (0.0-2.0)
            max_tokens: Maximum tokens to generate
            **kwargs: Additional service-specific options

        Yields:
            Message chunks as they arrive from the service. Each chunk contains:
            - role: "agent"
            - content: Partial text (may be a single token or character)
            - metadata: {"streaming": True, "model": model_name, "provider": provider_name}

        Raises:
            openai.APIError: For service-specific errors

        Example:
            >>> messages = [Message(role="user", content="Count to 10")]
            >>> async for chunk in llm.stream(messages):
            ...     print(chunk.content, end="", flush=True)
            1, 2, 3, 4, 5, 6, 7, 8, 9, 10

        Example - accumulate full response:
            >>> chunks = []
            >>> async for chunk in llm.stream(messages):
            ...     chunks.append(chunk.content)
            >>> full_response = "".join(chunks)

        Note:
            Not all OpenAI-compatible services support streaming. If the service
            doesn't support it, you'll get an error from the underlying service.
        """
        # Validate parameters
        self._validate_llm_params(temperature, max_tokens, **kwargs)

        # Convert messages
        openai_messages = self._convert_messages(messages)

        # Stream from OpenAI-compatible API
        stream = await self._client.chat.completions.create(
            model=self._model,
            messages=openai_messages,
            temperature=temperature,
            max_tokens=max_tokens,
            stream=True,
            **kwargs,
        )

        async for chunk in stream:
            # Extract content from delta
            delta = chunk.choices[0].delta
            if delta.content:
                yield Message(
                    role="agent",
                    content=delta.content,
                    metadata={
                        "streaming": True,
                        "model": self._model,
                        "provider": self._provider or "openai_compatible",
                    },
                )

    def _convert_messages(self, messages: list[Message]) -> list[dict[str, str]]:
        """
        Convert Agenkit Messages to OpenAI format.

        OpenAI-compatible services expect messages in the format:
        - role: "system", "user", or "assistant"
        - content: str

        Agenkit uses "agent" role which gets mapped to "assistant" for compatibility.

        Args:
            messages: List of Agenkit Messages

        Returns:
            List of message dicts in OpenAI format

        Note:
            The "tool" role is preserved as-is since many services support it.
        """
        openai_messages = []
        for msg in messages:
            # Map roles for OpenAI compatibility
            if msg.role in ("system", "user", "tool"):
                role = msg.role
            else:
                # Map "agent" and others to "assistant"
                role = "assistant"

            openai_messages.append({"role": role, "content": str(msg.content)})

        return openai_messages

    def unwrap(self) -> AsyncOpenAI:
        """
        Get the underlying AsyncOpenAI client.

        This provides an escape hatch for accessing OpenAI-specific features
        not exposed by the minimal LLM interface. Useful for advanced use cases
        like custom request handling or service-specific features.

        Returns:
            The AsyncOpenAI client configured with custom base_url

        Example:
            >>> llm = OpenAICompatibleLLM(
            ...     base_url="http://localhost:8000/v1",
            ...     model="llama-2-7b"
            ... )
            >>> client = llm.unwrap()
            >>> # Use OpenAI SDK features directly
            >>> response = await client.chat.completions.create(
            ...     model="llama-2-7b",
            ...     messages=[{"role": "user", "content": "Hello"}],
            ...     stream=True
            ... )

        Warning:
            Using unwrap() breaks provider portability. Code that uses unwrap()
            will need changes when switching between services or providers.
        """
        return self._client

File: agenkit-ts/src/llm/openai-compatible.ts
/**
 * OpenAI-Compatible LLM adapter.
 *
 * Generic adapter for OpenAI-compatible inference services like vLLM,
 * llama.cpp, SGLang, TensorRT-LLM, and others.
 *
 * This adapter enables Agenkit to work with any service implementing the
 * OpenAI Chat Completions API by configuring the OpenAI SDK with a custom
 * base URL.
 *
 * Supported services:
 * - vLLM: High-throughput batch inference
 * - llama.cpp: Lightweight C++ implementation (CPU-friendly)
 * - SGLang: Optimized for complex prompts
 * - TensorRT-LLM: NVIDIA GPU optimized
 * - OpenLLM: Multi-model serving platform
 * - MLC LLM: Mobile and edge deployment
 * - Text Generation Inference (TGI): HuggingFace inference server
 * - Inferflow: High-performance inference
 *
 * Example - vLLM local deployment:
 *
 *   const agent = new OpenAICompatibleAgent({
 *     baseURL: 'http://localhost:8000/v1',
 *     model: 'meta-llama/Llama-2-7b-chat-hf',
 *     provider: 'vllm',
 *   });
 *
 *   const response = await agent.process({
 *     role: 'user',
 *     content: 'Hello!',
 *   });
 *
 * Example - llama.cpp server:
 *
 *   const agent = new OpenAICompatibleAgent({
 *     baseURL: 'http://localhost:8080/v1',
 *     model: 'llama-2-7b-chat',
 *     provider: 'llamacpp',
 *   });
 *
 * Example - Streaming:
 *
 *   for await (const chunk of agent.processStream(message)) {
 *     process.stdout.write(chunk.content as string);
 *   }
 */

import OpenAI from 'openai';
import { Agent, Message, createMessage } from '../core/interfaces.js';

/**
 * OpenAI-Compatible adapter configuration.
 */
export interface OpenAICompatibleConfig {
  /**
   * Base URL of the inference service (e.g., "http://localhost:8000/v1").
   * Must include the /v1 suffix for most services.
   */
  baseURL: string;

  /**
   * Model name/identifier used by the inference service.
   * Format varies by service:
   * - vLLM: "meta-llama/Llama-2-7b-chat-hf"
   * - llama.cpp: "llama-2-7b-chat"
   * - SGLang: "meta-llama/Llama-2-13b-chat-hf"
   */
  model: string;

  /**
   * Optional provider name for metadata and debugging.
   * Examples: "vllm", "llamacpp", "sglang", "tensorrt"
   */
  provider?: string;

  /**
   * Optional API key. Most local services don't require authentication,
   * so this can be omitted. Defaults to "not-needed".
   */
  apiKey?: string;

  /** Agent name (default: derived from provider or "openai-compatible") */
  name?: string;

  /** Temperature (default: 0.7) */
  temperature?: number;

  /** Max tokens (default: 1024) */
  maxTokens?: number;

  /** Top P (default: 1.0) */
  topP?: number;

  /** Request timeout in milliseconds (default: 60000) */
  timeout?: number;

  /** Additional OpenAI client options */
  options?: Partial<OpenAI.Chat.ChatCompletionCreateParams>;
}

/**
 * OpenAICompatibleAgent implements Agent interface for OpenAI-compatible services.
 *
 * This adapter provides a consistent interface across different local and
 * self-hosted inference engines by using the OpenAI SDK with a custom base URL.
 *
 * Features:
 * - Works with 8+ OpenAI-compatible inference services
 * - Streaming support
 * - Configurable parameters (temperature, max_tokens, etc.)
 * - Provider metadata for debugging and monitoring
 * - Automatic message format conversion
 *
 * Usage:
 *
 *   // vLLM local deployment
 *   const agent = new OpenAICompatibleAgent({
 *     baseURL: 'http://localhost:8000/v1',
 *     model: 'meta-llama/Llama-2-7b-chat-hf',
 *     provider: 'vllm',
 *   });
 *
 *   // Basic completion
 *   const response = await agent.process({
 *     role: 'user',
 *     content: 'What is machine learning?',
 *   });
 *   console.log(response.content);
 *
 *   // Streaming
 *   for await (const chunk of agent.processStream(message)) {
 *     process.stdout.write(chunk.content as string);
 *   }
 *
 *   // Access metadata
 *   console.log(response.metadata?.provider);  // "vllm"
 *   console.log(response.metadata?.baseURL);   // "http://localhost:8000/v1"
 */
export class OpenAICompatibleAgent implements Agent {
  readonly name: string;
  readonly capabilities: string[];

  private client: OpenAI;
  private model: string;
  private provider: string;
  private baseURL: string;
  private temperature: number;
  private maxTokens: number;
  private topP: number;
  private options: Partial<OpenAI.Chat.ChatCompletionCreateParams>;

  constructor(config: OpenAICompatibleConfig) {
    // Set provider and name
    this.provider = config.provider || 'openai_compatible';
    this.name = config.name || this.provider;

    // Initialize OpenAI client with custom base URL
    this.client = new OpenAI({
      baseURL: config.baseURL,
      apiKey: config.apiKey || 'not-needed',
      timeout: config.timeout || 60000,
    });

    this.model = config.model;
    this.baseURL = config.baseURL;
    this.temperature = config.temperature ?? 0.7;
    this.maxTokens = config.maxTokens || 1024;
    this.topP = config.topP ?? 1.0;
    this.options = config.options || {};

    // Set capabilities based on provider
    this.capabilities = [
      'openai-compatible',
      'chat',
      'streaming',
      this.provider,
    ];
  }

  /**
   * Convert agenkit Message to OpenAI message format.
   *
   * OpenAI-compatible services expect messages in the format:
   * - role: "system", "user", or "assistant"
   * - content: string
   *
   * Agenkit uses generic roles which get mapped to OpenAI format.
   */
  private toOpenAIMessage(
    message: Message,
  ): OpenAI.Chat.ChatCompletionMessageParam {
    const content =
      typeof message.content === 'string'
        ? message.content
        : JSON.stringify(message.content);

    // Map roles for OpenAI compatibility
    let role: 'system' | 'user' | 'assistant';
    if (message.role === 'system') {
      role = 'system';
    } else if (message.role === 'assistant' || message.role === 'agent') {
      role = 'assistant';
    } else {
      role = 'user';
    }

    return { role, content };
  }

  /**
   * Process a message using the OpenAI-compatible service.
   *
   * @param message Input message
   * @returns Response message with metadata including:
   *   - model: Model identifier used
   *   - usage: Token counts (prompt_tokens, completion_tokens, total_tokens)
   *   - finish_reason: Why generation stopped
   *   - provider: Provider name for debugging
   *   - baseURL: Service URL for debugging
   *   - id: Response ID from the service
   */
  async process(message: Message): Promise<Message> {
    const openaiMessage = this.toOpenAIMessage(message);

    // Spreading `...this.options` defeats TS's literal-narrowing on
    // `stream: false`, so the result is typed as the streaming/non-streaming
    // union. We pass stream:false, so assert the non-streaming response.
    const response = (await this.client.chat.completions.create({
      model: this.model,
      messages: [openaiMessage],
      temperature: this.temperature,
      max_tokens: this.maxTokens,
      top_p: this.topP,
      ...this.options,
      stream: false,
    })) as OpenAI.Chat.ChatCompletion;

    // Check for valid response
    if (!response.choices || response.choices.length === 0) {
      throw new Error('Service returned no choices');
    }

    const choice = response.choices[0];
    const content = choice.message.content || '';

    return createMessage('assistant', content, {
      model: response.model,
      usage: {
        prompt_tokens: response.usage?.prompt_tokens || 0,
        completion_tokens: response.usage?.completion_tokens || 0,
        total_tokens: response.usage?.total_tokens || 0,
      },
      finish_reason: choice.finish_reason,
      id: response.id,
      provider: this.provider,
      baseURL: this.baseURL,
    });
  }

  /**
   * Process a message with streaming response.
   *
   * This method streams response chunks as they're generated by the service,
   * enabling real-time display and lower perceived latency.
   *
   * @param message Input message
   * @returns Async iterator of response chunks
   *
   * Example:
   *
   *   const message = { role: 'user', content: 'Count to 10' };
   *   for await (const chunk of agent.processStream(message)) {
   *     process.stdout.write(chunk.content as string);
   *   }
   *
   * Note: Not all OpenAI-compatible services support streaming. If the service
   * doesn't support it, you'll get an error from the underlying service.
   */
  async *processStream(
    message: Message,
  ): AsyncGenerator<Message, void, undefined> {
    const openaiMessage = this.toOpenAIMessage(message);

    // See process(): the options spread defeats stream-literal narrowing, so
    // assert the streaming response type (we pass stream:true).
    const stream = (await this.client.chat.completions.create({
      model: this.model,
      messages: [openaiMessage],
      temperature: this.temperature,
      max_tokens: this.maxTokens,
      top_p: this.topP,
      ...this.options,
      stream: true,
    })) as AsyncIterable<OpenAI.Chat.ChatCompletionChunk>;

    for await (const chunk of stream) {
      const delta = chunk.choices[0]?.delta;
      if (delta?.content) {
        yield createMessage('assistant', delta.content, {
          model: this.model,
          streaming: true,
          provider: this.provider,
        });
      }
    }
  }

  /**
   * Get the underlying OpenAI client.
   *
   * This provides an escape hatch for accessing OpenAI SDK features
   * not exposed by the minimal Agent interface.
   *
   * Warning: Using getClient() breaks provider portability. Code that uses
   * getClient() will need changes when switching between services.
   */
  getClient(): OpenAI {
    return this.client;
  }
}
